Added IconThemePlugin

ajmeese7 opened the pull request:
This is intended to update the metadata of a package to include the icon file types, allowing for a mix-and-match of `svg`, `png`, `gif`, etc.

This has not been tested in a `dev-meta` context, but when it is included in a local repo, the usage is as follows:

```js
// In the `webpack.config.js` file
const IconThemePlugin = require('./iconThemePlugin');

...

plugins: [
  new IconThemePlugin(),
  ...plugins,
],
```

I assume that with this package, importing will be something along the lines of:

```js
import IconThemePlugin from '@osjs/dev-meta/IconThemePlugin';
```

But I have not had a chance to set up a local OS.js env and test yet.

----

Related to https://github.com/os-js/osjs-client/pull/191.

ajmeese7 committed:
Added IconThemePlugin

andersevenrud commented on IconThemePlugin.js:
```diff
@@ -0,0 +1,37 @@
+
+      // Create an object with the property name as the asset name
+      // and the value as the file extension
+      const iconTypes = icons.reduce(
```
I find that `reduce` adds quite a bit of cognitive load.

You can write this as the following:

```js
const iconEntries = icons.map(icon => icon.split('.')) // NB: This does not work as you'd expect if someone puts dots in their names (same with your original implementation):)
const iconTypes = Object.fromEntries(iconEntries)
```

ajmeese7 replied:
I can easily mod the `reduce` to use `pop()` on the split to get the extension:

```js
const iconTypes = icons.reduce(
	(acc, icon) => {
		const split = icon.split('.');
		const ext = split.pop();
		const name = split.join('.');
		acc[name] = ext;
		return acc;
	},
	{}
);
```

But I'm not sure if there is an easy way to do so with `map`. Any suggestions?

andersevenrud replied:
Suppose it could be done like this:

```js
const iconEntries = icons.map(file => [
  file.substr(0, file.lastIndexOf('.')),
  file.substr(file.lastIndexOf('.') + 1, file.length)
])

const iconTypes = Object.fromEntries(iconEntries)
```

ajmeese7 replied:
Whoa I had no idea that `Object.fromEntries` would work with an array of arrays like that, cool stuff! Good idea.

andersevenrud commented on IconThemePlugin.js:
```diff
@@ -0,0 +1,37 @@
+        {}
+      );
+
+      // Read the current metadata file
```
I don't think the comments here are needed. It's pretty self explanatory, especially if the above reduce is replaced.

ajmeese7 replied:
No problem, will remove 👍🏼

andersevenrud replied:
A nicer way (imo) to handle documentation of stuff like this is just to add a docblock in the top of the file that explains what the module does on a technical level.

Since this uses APIs that are well documented and whatnot, there should not be an issue in understanding what's going on as it's easy to look up :)

andersevenrud reviewed (changes requested):
Seems like you accidentally included a `package-lock` update here. Revert this.

Also the `.gitignore` really should be a separate commit (or a separate PR).

andersevenrud commented:
> But I have not had a chance to set up a local OS.js env and test yet.

It's a shame that linking meta packages like this basically doesn't work in npm anymore :(

A hacky workaround is to just copy the `dev-meta` local source into `node_modules` when you need it for stuff like this. 

Some package managers allow for this in a monorepo setup though, but linking is a no-go.

> I assume that with this package, importing will be something along the lines of:

Indeed.

ajmeese7 commented:
> Seems like you accidentally included a package-lock update here. Revert this.
> Also the .gitignore really should be a separate commit (or a separate PR).

The `package-lock.json` was actually intentional, since the lock file is currently out of date as compared to the `package.json`. But no problem, both changes can be removed and added in a separate PR.

> It's a shame that linking meta packages like this basically doesn't work in npm anymore :(

100% agreed, that's the main reason I transitioned to `rush` for my monorepo in the first place, because it was so impossible to locally test all `dev-meta` updates. I got frustrated trying to update to Webpack 5 since it is currently so difficult to configure.

andersevenrud commented:
> The package-lock.json was actually intentional, since the lock file is currently out of date as compared to the package.json.

This is actually perfectly fine. It's only metadata that is not in sync. This file is not included for distributed packages, nor does it have any impact on local development (git checkouts) :)

ajmeese7 commented:
Yep, just checking out my Codespaces automatically made the account and the `.gitignore` since that's part of their setup process, so I figured it wouldn't hurt to push the change. I often forget that other people organize their commits a lot more, I usually push all the working code that I have changed even if it is not specifically related to the purpose of the commit.

Slowly working on it though :)

ajmeese7 committed:
Made requested changes

andersevenrud commented:
> I often forget that other people organize their commits a lot more

I touch on this quite a lot because of the following benefits:

* Obvious what happened by scanning git logs
* Makes the "blame" feature work as it was intended
* Takes away so much pain when it comes to merges and doing bisects if something goes wrong down the line, because:
  * You can revert a single commit instead of a large one + go back and add in what you wanted to keep
  * If a collision occurs you get the advantages of the first two points
  * And when doing bisects the above also applies

I use https://www.conventionalcommits.org/en/v1.0.0/ pretty much in all projects nowadays, but never adapted that in this one because it was started way before I knew about that. But I'm considering enforcing that as well. So easy to automate semver and changelogs that way. And it also goes into the whole thinking about stuff in smaller chunks.

andersevenrud commented on IconThemePlugin.js:
```diff
@@ -0,0 +1,26 @@
+      const metadata = require('./metadata.json');
+      metadata.icons = iconTypes;
+      const json = JSON.stringify(metadata, null, 2);
+      compilation.emitAsset('../metadata.json', new RawSource(json));
```
I'm not entirely sure it's completely safe to hard-code this one. An edge-case I can think of is that someone changes their dist to be something deep.

As mentioned in an earlier thread, maybe it's a way to detect the root directory ?

ajmeese7 replied:
I'm not seeing a good way to detect the project root from within a Webpack plugin; how about if I add class options that give the user the ability to override the path in a parameter if they choose?

I *think* that's what you were suggesting in the next comment, but I am not sure.

andersevenrud replied:
> I'm not seeing a good way to detect the project root from within a Webpack plugin;

Hm. I vaguely remember being able to grab the path to the physical `webpack.config.<ext>` file from within a plugin, then just do a dirname on it to extract the root. Even if that memory is false, I would imagine that there is a function or variable somewhere to get that.

> I think that's what you were suggesting in the next comment, but I am not sure.

Yes.

Which would just default to the detected path, if possible.

ajmeese7 replied:
The functions I saw to accomplish that all required changes to the actual `webpack.config.js` to enable it, but I will keep looking to make sure I haven't missed something obvious. If I can't find anything I'll push the constructor with options and we can go from there.

ajmeese7 replied:
I found a weird way to do it, since I didn't see any easy officially documented methods. I used `compilation.compiler.context` to get the project root, and it worked when the output dir was normal and when I set the output to `/tmp` for testing, so I think it works as we want it to.

ajmeese7 replied:
Found a fun problem with it that I've been working to resolve, `compilation.emitAsset` doesn't accept an absolute path. Got it working by generating a relative path from the compilation output path to the context path via the `path.relative` method. Going to push a commit shortly.

andersevenrud commented on IconThemePlugin.js:
```diff
@@ -0,0 +1,26 @@
+      ]);
+      const iconTypes = Object.fromEntries(iconEntries);
+
+      const metadata = require('./metadata.json');
```
This also relates to what I just reviewed. It is possible that someone has automation to generate stuff. So an option as discussed previously might be a solution for that :thinking:

ajmeese7 committed:
Dynamically determine metadata file path

ajmeese7 commented:
If you're interested, I have also modified my instance to alphabetize the icon names and filter out `.map` and `.js` file types, for a cleaner end result. The order is not always the same for the assets, so sorting them by name ensures that there are no arbitrary metadata file changes without any updates to the underlying content. And we are obviously not interested in those non-icon file types in the metadata, so I removed the only instances of such that I have encountered, which are `.map` and `.js` files.

andersevenrud commented:
> filter out .map and .js file types, for a cleaner end result. 

Ideally this implementation should pick *only* the icons.

> no arbitrary metadata file changes without any updates to the underlying content

There's a handy term for this: Idempotence ! I think this should be implemented as well.

andersevenrud reviewed (changes requested):
There's still the case of a missing plugin option to allow customizing the path as previously discussed.

Assuming that everyone uses the standard build setup is a bit risky, and I know of a few that has their own setup that builds all packages and the core in one setup that involves custom paths etc.

So some way of controlling the root of `metadata.json`  (that overrides `compiler.context`) must be added.

ajmeese7 committed:
Made requested changes

andersevenrud commented on IconThemePlugin.js:
```diff
@@ -0,0 +1,44 @@
+
+  apply(compiler) {
+    const pluginName = IconThemePlugin.name;
+    const {webpack} = compiler;
```
Just a small tip. It's possible to do deep destructuring :raised_hands: 

```
const {webpack: {RawResource}} = compiler
```

ajmeese7 replied:
Assuming you mean `RawSource`, correct?

andersevenrud replied:
Yes. I read that wrong.

andersevenrud commented on IconThemePlugin.js:
```diff
@@ -0,0 +1,44 @@
+        file.substr(file.lastIndexOf('.') + 1, file.length)
+      ]);
+
+      const sortedIcons = Object.fromEntries(
```
Technically, you don't need to do this, because `.sort()` actually mutates the array (yuck... I know).

So

```
iconEntries.sort((a, b) => a[0].localeCompare(b[0]))
```

And then 

```
      metadata.icons = Object.fromEntries(
        Object.entries(iconEntries).filter(
          ([key, value]) => imageTypes.includes(value)
        )
      );
```

Gets you the same result, just a little bit less code :)

andersevenrud replied:
Also, this might be a bit nicer:

```
iconEntries.sort(([a], [b]) => a.localeCompare(b))
```

ajmeese7 committed:
Implemented suggested changes

andersevenrud reviewed (commented):
LGTM

However, I just thought of an edge-case here. The detection here assumes all lower-case file extensions.

Could this cause an issue with Windows and Mac environments where this is all over the place ?

Edit: Nah.... let's not go there. This is only an edge-case for icon names that falls outside the standard naming conventions (GTK). Let's just make it a rule to use lower case for all resources :)

andersevenrud commented:
Thanks!

Squashed and re-worded commit for more context

andersevenrud commented:
Published with 2.3.0
